### PR TITLE
API naming for Expression

### PR DESCRIPTION
**Pull request by Danieladu**
link: #3443
API naming for Expression
1. Rename ExpressionEngine to ExpressionParser
2. Rename ExpresionLexer.g4 to ExpressionAntlrLexer.g4 and rename ExpressionParser.g4 to ExpressionAntlrParser.g4 to avoid conflict
3. Restruct function scope

### CODE REVIEW

File: tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/LGFileTest.cs [hunk 1]
```diff
@@ -988,7 +988,7 @@ public void TestInlineEvaluate()
         [TestMethod]
         public void TestCustomFunction()
         {
-            var engine = new ExpressionEngine((string func) =>
+            var engine = new ExpressionParser((string func) =>
```
Comment (boydc2014) [inline tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/LGFileTest.cs]: s/engine/parser

File: libraries/AdaptiveExpressions/Expression.cs [hunk 2]
```diff
@@ -104,7 +104,7 @@ public Expression(ExpressionEvaluator evaluator, params Expression[] children)
         /// allow a string to be implicitly assigned to an expression property.
         /// </summary>
         /// <param name="expression">string expression.</param>
-        public static implicit operator Expression(string expression) => Expression.Parse(expression);
+        public static implicit operator Expression(string expression) => Parse(expression);
```
Comment (boydc2014) [inline libraries/AdaptiveExpressions/Expression.cs]: what is this for?

File: libraries/AdaptiveExpressions/Expression.cs [hunk 3]
```diff
@@ -113,158 +113,9 @@ public Expression(ExpressionEvaluator evaluator, params Expression[] children)
         /// <returns>resulting expression.</returns>
         public static Expression Parse(string expression)
         {
-            return new ExpressionEngine().Parse(expression);
+            return new ExpressionParser().Parse(expression);
```
Comment (boydc2014) [inline libraries/AdaptiveExpressions/Expression.cs]: why not moving this also into expression factory

File: tests/Microsoft.Bot.Builder.TestBot.Json/CustomSteps/MultiplyDialog.cs [hunk 4]
```diff
@@ -50,8 +50,8 @@ public MultiplyDialog([CallerFilePath] string sourceFilePath = "", [CallerLineNu
 
         public override Task<DialogTurnResult> BeginDialogAsync(DialogContext dc, object options = null, CancellationToken cancellationToken = default(CancellationToken))
         {
-            var (arg1, err1) = Expression.Parse(Arg1).TryEvaluate(dc.GetState());
-            var (arg2, err2) = Expression.Parse(Arg2).TryEvaluate(dc.GetState());
+            var (arg1, err1) = new ExpressionParser().Parse(Arg1).TryEvaluate(dc.GetState());
```
Comment (tomlm) [inline tests/Microsoft.Bot.Builder.TestBot.Json/CustomSteps/MultiplyDialog.cs]: >ExpressionParser [](start = 35, length = 16)

NO why is this happening? 

Expression.Parse was there for a reason.

File: libraries/AdaptiveExpressions/Expression.cs [hunk 5]
```diff
@@ -104,166 +104,7 @@ public Expression(ExpressionEvaluator evaluator, params Expression[] children)
...
-        /// Construct and validate a property accessor.
-        /// </summary>
-        /// <param name="property">Property to lookup.</param>
-        /// <param name="instance">Expression to get instance that contains property or null for global state.</param>
-        /// <returns>New expression.</returns>
-        public static Expression Accessor(string property, Expression instance = null)
-            => instance == null
-            ? MakeExpression(ExpressionType.Accessor, ConstantExpression(property))
-            : MakeExpression(ExpressionType.Accessor, ConstantExpression(property), instance);
+        public static implicit operator Expression(string expression) => new ExpressionParser().Parse(expression);
```
Comment (tomlm) [inline libraries/AdaptiveExpressions/Expression.cs]: Why was this removed?  I far far prefer this signature!

File: tests/Microsoft.Bot.Builder.TestBot.Json/CustomSteps/MultiplyDialog.cs (hunk 4 above)
Comment (tomlm) [inline tests/Microsoft.Bot.Builder.TestBot.Json/CustomSteps/MultiplyDialog.cs]: >new ExpressionParser( [](start = 31, length = 21)

ExpressionBuilder.Parse() please

**Comment by vishwacsena** [comment]
I chatted with @tomlm; we should remove ExpressionBuilder and move those static back to Expression so we reduce the concept count here and eliminate the need for additional class. 

@Danieladu @boydc2014

File: libraries/AdaptiveExpressions/Expression.cs (hunk 3 above)
Comment (tomlm) [reply libraries/AdaptiveExpressions/Expression.cs]: NO, I think it is fewer concept count to have it be on expression. 
Expression.And(Expression.Lamda(() => ...), Expression.Parse("user.age > 3"))
is WAY better then
ExpressionBuilder.And(ExpressionBuilder.Lamda(() => ...), ExpressionBuilder.Parse("user.age > 3))

Only one class, and it's all about Expressions and discoverable off of expression.


---
In reply to: [386273952](https://github.com/microsoft/botbuilder-dotnet/pull/3471#discussion_r386273952) [](ancestors = 386273952)

File: libraries/AdaptiveExpressions/Expression.cs (hunk 5 above)
Comment (Danieladu) [reply libraries/AdaptiveExpressions/Expression.cs]: just moved it to a new class "ExpressionBuilder"

File: tests/Microsoft.Bot.Builder.TestBot.Json/CustomSteps/MultiplyDialog.cs (hunk 4 above)
Comment (boydc2014) [reply tests/Microsoft.Bot.Builder.TestBot.Json/CustomSteps/MultiplyDialog.cs]: is expressionPraser.parse more consistent lgParser.parse?

**Comment by boydc2014** [comment]
> I chatted with @tomlm; we should remove ExpressionBuilder and move those static back to Expression so we reduce the concept count here and eliminate the need for additional class.
> 
> @Danieladu @boydc2014

Sure, i'm less opinionated on moving static method back.  I do prefer ExpressionParser than ExpressionBuilder\Engine\xxx, because it aligns very well with LGParser.parser, and it feels very natural. One indicator is that, in today's sdk, we have server places that says "ExpressionEngine parser", indicating that people (maybe @tomlm you wrote this :) ) that it's parser.  make sense?

File: tests/Microsoft.Bot.Builder.TestBot.Json/CustomSteps/MultiplyDialog.cs (hunk 4 above)
Comment (tomlm) [reply tests/Microsoft.Bot.Builder.TestBot.Json/CustomSteps/MultiplyDialog.cs]: No, I prefer Expression.Parse(Arg1).TryEvaluate()
It just makes the concept count lower if you are looking at this as a sample

---
In reply to: [386753486](https://github.com/microsoft/botbuilder-dotnet/pull/3471#discussion_r386753486) [](ancestors = 386753486)

File: libraries/AdaptiveExpressions/ExpressionProperties/ExpressionProperty.cs [hunk 6]
```diff
@@ -70,19 +70,19 @@ public virtual Expression ToExpression()
 
             if (this.ExpressionText != null)
             {
-                this.expression = Expression.Parse(this.ExpressionText.TrimStart('='));
+                this.expression = new ExpressionParser().Parse(this.ExpressionText.TrimStart('='));
```
Comment (tomlm) [inline libraries/AdaptiveExpressions/ExpressionProperties/ExpressionProperty.cs]: >new ExpressionParser() [](start = 34, length = 22)

Please put Expression.Parse back across these files.  It reduces the concept count. (You only need to think about Expression class unless you want to understand the backing stuff.)

Comment (boydc2014) [reply libraries/AdaptiveExpressions/ExpressionProperties/ExpressionProperty.cs]: yes, make sense, let's do that
